### PR TITLE
Simplifications in match lowering

### PR DESCRIPTION
**Pull request by Nadrieril**
A series of small simplifications and deduplications in the MIR lowering of patterns.

r? @matthewjasper

### CODE REVIEW

File: compiler/rustc_mir_build/src/build/matches/mod.rs [hunk 1]
```diff
@@ -725,55 +726,49 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
...
                         *match_place = Some(place);
-                    }
-                }
-                // All of the subcandidates should bind the same locals, so we
-                // only visit the first one.
-                next = candidate_ref.subcandidates.get(0)
+                    } else {
+                        bug!("Let binding to non-user variable.")
+                    };
+                });
```
Comment (Nadrieril) [inline compiler/rustc_mir_build/src/build/matches/mod.rs:686]: I guess this isn't so much a simplification by itself; it's just making my life simpler as I try to clarify this whole thing.

File: compiler/rustc_mir_build/src/build/block.rs [hunk 2]
```diff
@@ -189,38 +189,37 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
...
+                        let (matching, failure) =
+                            this.in_if_then_scope(last_remainder_scope, else_block_span, |this| {
+                                this.lower_let_expr(
+                                    block,
+                                    *initializer,
+                                    pattern,
+                                    None,
+                                    initializer_span,
+                                    false,
+                                    true,
```
Comment (Zalathar) [inline compiler/rustc_mir_build/src/build/block.rs:217]: This sort of thing makes me want to start adding dedicated enums to replace some of these magic booleans.

For example, I want to replace `declare_let_bindings: bool` with a three-way enum:
```rust
enum LetBindingsMode {
    Normal,
    Inhibit,
    LetNotPermitted,
}
```

Comment (Zalathar) [reply compiler/rustc_mir_build/src/build/block.rs:217]: I've opened a draft PR to replace some booleans with enums: #126981.

**Comment by matthewjasper** [comment]
I think it makes sense to merge this separately from #126981.

@bors r+

**Pull request merged**